b2share/modules/deposit/api.py

```diff
@@ -252,6 +265,35 @@ def commit(self):
         This method extends the default implementation by publishing the
         deposition when 'publication_state' is set to 'published'.
         """
+        if 'external_pids' in self:
+            deposit_id = self['_deposit']['id']
+            recid = PersistentIdentifier.query.filter_by(
+                pid_value=deposit_id).first()
+            assert recid.status == 'R'
+            record_bucket = RecordsBuckets.query.filter_by(
+                record_id=recid.pid_value).first()
+            bucket = Bucket.query.filter_by(id=record_bucket.bucket_id).first()
+            object_versions = ObjectVersion.query.filter_by(
+                bucket_id=bucket.id).all()
+            # for the existing files
+            for object_version in object_versions:
+                if object_version.file is None or \
+                        object_version.file.storage_class != 'B':
+                    continue
+                # check that they are still in the file pids list or remove
+                if object_version.key not in self['external_pids']:
+                    ObjectVersion.delete(bucket,
+                                         object_version.key)
+                # check that the uri is still the same or update it
+                elif object_version.file.uri != \
+                        self['external_pids'][object_version.key]:
+                    db.session.query(FileInstance).\
+                        filter_by(key=object_version.key).\
+                        update({"uri": self['external_pids'][object_version.key]})
+            create_b2safe_file(self['external_pids'], bucket)
+            self['_deposit']['external_pids'] = self['external_pids']
```

Comment:
**This is dangerous. This stores anything the user sends without any structure constraint. It will eventually break the indexing when the field types conflict.**